File: CelebrityCrunch/CelebrityCrunch/RWTActor.m
//
//  RWTActor.m
//  CookieCrunch
//
//  Created by DJIBRIL KEITA on 10/8/14.
//  Copyright (c) 2014 DJIBRILKEITA. All rights reserved.
//

#import "RWTActor.h"

// Define NumActors based on the count from JSON, or keep it fixed if JSON is guaranteed to have this many.
// For now, we'll keep the extern declaration and define it after loading.
const NSUInteger NumActors = 70; // Default, will be updated if JSON loading is successful and count differs.

@interface RWTActor ()

@end

// Static variables to hold loaded actor data
static NSArray *_actorDefinitions = nil;
static NSDictionary *_actorDefinitionsByName = nil;
static dispatch_once_t _loadActorsOnceToken;


@implementation RWTActor

+ (void)loadActorDefinitions {
    dispatch_once(&_loadActorsOnceToken, ^{
        NSString *path = [[NSBundle mainBundle] pathForResource:@"Actors" ofType:@"json"];
        if (path == nil) {
            NSLog(@"Could not find Actors.json");
            _actorDefinitions = @[]; // Empty array to prevent further load attempts
            _actorDefinitionsByName = @{};
            return;
        }
        
        NSError *error;
        NSData *data = [NSData dataWithContentsOfFile:path options:0 error:&error];
        if (data == nil) {
            NSLog(@"Could not load data from Actors.json: %@", error.localizedDescription);
            _actorDefinitions = @[];
            _actorDefinitionsByName = @{};
            return;
        }
        
        NSArray *jsonArray = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
        if (jsonArray == nil || ![jsonArray isKindOfClass:[NSArray class]]) {
            NSLog(@"Could not parse Actors.json: %@", error.localizedDescription);
            _actorDefinitions = @[];
            _actorDefinitionsByName = @{};
            return;
        }
        
        _actorDefinitions = jsonArray;
        NSMutableDictionary *byName = [NSMutableDictionary dictionaryWithCapacity:[_actorDefinitions count]];
        for (NSDictionary *actorDict in _actorDefinitions) {
            NSString *name = actorDict[@"name"];
            if (name) {
                [byName setObject:actorDict forKey:name];
            }
        }
        _actorDefinitionsByName = [byName copy];
        
        // Optionally update NumActors if it should be dynamic
        // NumActors = [_actorDefinitions count]; // Uncomment if NumActors should reflect actual loaded count
        if ([_actorDefinitions count] != NumActors) {
             NSLog(@"Warning: NumActors constant (%lu) does not match loaded actor definitions count (%lu).", (unsigned long)NumActors, (unsigned long)[_actorDefinitions count]);
        }
    });
}

+ (NSDictionary *)actorDataForIndex:(NSInteger)index { // index is 1-based
    [RWTActor loadActorDefinitions];
    if (index > 0 && index <= [_actorDefinitions count]) {
        return _actorDefinitions[index - 1]; // Convert 1-based to 0-based for array access
    }
    NSLog(@"Error: actorDataForIndex: index %ld out of bounds (1-%lu).", (long)index, (unsigned long)[_actorDefinitions count]);
    return nil;
}

+ (NSInteger)actorIndexForName:(NSString *)name { // returns 1-based index
    [RWTActor loadActorDefinitions];
    NSDictionary *actorDict = _actorDefinitionsByName[name];
    if (actorDict) {
        return [actorDict[@"id"] integerValue];
    }
    return NSNotFound; // Or 0 if that's the convention for not found
}

-(NSString *)spriteName {
    NSDictionary *data = [RWTActor actorDataForIndex:self.actorIndex];
    return data[@"spriteBase"]; // Assuming spriteBase is the direct name for SKTexture
}

-(NSString *)highlightedSpriteName {
    NSDictionary *data = [RWTActor actorDataForIndex:self.actorIndex];
    if (data && data[@"spriteBase"] && data[@"highlightedSuffix"]) {
        return [NSString stringWithFormat:@"%@%@", data[@"spriteBase"], data[@"highlightedSuffix"]];
    }
    return nil; // Or a default/error sprite name
}

// Corrected typo from blueHilightedSpriteNames to blueHighlightedSpriteName
-(NSString *)blueHighlightedSpriteName {
    NSDictionary *data = [RWTActor actorDataForIndex:self.actorIndex];
     if (data && data[@"spriteBase"] && data[@"blueSuffix"]) {
        return [NSString stringWithFormat:@"%@%@", data[@"spriteBase"], data[@"blueSuffix"]];
    }
    return nil; // Or a default/error sprite name
}

// Instance method uses the new class method
-(NSUInteger)actorIndex:(NSString *)actorName {
    NSInteger index = [RWTActor actorIndexForName:actorName];
    if (index == NSNotFound) {
        // Handle NSNotFound appropriately, perhaps return 0 or a specific error value
        // For now, returning 0 to match previous behavior if actorName was not in the static array.
        // However, the original returned index (0 to 69). actorIndexForName returns 1-70 or NSNotFound.
        // The caller in RWTLevel expects 0-based index from this method.
        // Let's adjust to return 0-based index or NSNotFound (which is a large unsigned value).
        // If a 0-based index is strictly required by the caller, this needs careful mapping.
        // The problem description mentions: "NSUInteger actorSpriteIndex = [tempActorForIndex actorIndex:selectedActorName];"
        // and "index = actorSpriteIndex + 1;"
        // This means the original actorIndex: method returned a 0-based index.
        // So, if [RWTActor actorIndexForName:actorName] returns 1-based ID, we subtract 1.
        
        // If actorIndexForName returns 1-based ID (e.g., 1 for Robert-Redford)
        // and the old system returned 0 for "Robert-Redford", then we need to return id-1.
        // If NSNotFound, return NSNotFound or handle as error.
        return NSNotFound; // Or some other error indicator if 0 is a valid index
    }
    return index -1; // Convert 1-based ID from JSON to 0-based index
}


-(NSArray *)getMovieListForActor:(RWTActor*)actor withMap:(NSDictionary *)map{
    // This method might need to get the actor's name using the new system
    // if actor.actorName is not reliably set from outside.
    // Assuming actor.actorIndex is set, we can get the name:
    NSDictionary *actorData = [RWTActor actorDataForIndex:actor.actorIndex];
    NSString *actorNameFromData = actorData[@"name"];
    
    if (actorNameFromData) {
        return [map objectForKey:actorNameFromData];
    }
    // Fallback or error if name isn't found, though spriteName would also fail.
    // Original used [actor spriteName] which now uses the new system too.
    return [map objectForKey:[actor spriteName]]; // This should still work
}

@end

File: CelebrityCrunch/CelebrityCrunch/RWTActor.swift
import Foundation
import SpriteKit // For SKSpriteNode

// Using NSNotFound for consistency if this value is used in Obj-C parts that expect it.
// Otherwise, returning an optional Int? (nil for not found) would be more Swifty.
// public let NSNotFound: Int = -1 // Standard definition

@objc(RWTActor)
public class RWTActor: NSObject {

    @objc public var actorName: String = "" // May not be strictly needed if actorIndex is the primary ID
    @objc public var column: Int = 0
    @objc public var row: Int = 0
    @objc public var actorIndex: Int = 0 // 1-based index from JSON
    @objc public var sprite: SKSpriteNode?
    @objc public var machted: Bool = false // Typo for 'matched'

    // Default initializer
    public override init() {
        super.init()
        // Call loadActorDefinitions to ensure data is ready when an instance is created,
        // especially if actorName might be set based on actorIndex soon after.
        // However, instance methods like spriteName() will call it lazily if not.
        // RWTActor.loadActorDefinitions() // Not strictly necessary here due to lazy loading in accessors
    }
    
    // --- Static Data & JSON Loading ---

    public static let NumActors = 70 // Or derive from _actorDefinitions.count after loading

    // Private static properties for lazy loading JSON data
    private static let _actorDefinitionsStorage: [[String: Any]]? = {
        guard let path = Bundle.main.path(forResource: "Actors", ofType: "json") else {
            print("Could not find Actors.json")
            return []
        }
        do {
            let data = try Data(contentsOf: URL(fileURLWithPath: path), options: .mappedIfSafe)
            let jsonResult = try JSONSerialization.jsonObject(with: data, options: .mutableLeaves)
            if let jsonArray = jsonResult as? [[String: Any]] {
                if jsonArray.count != NumActors {
                    print("Warning: NumActors constant (\(NumActors)) does not match loaded actor definitions count (\(jsonArray.count)).")
                }
                return jsonArray
            }
        } catch {
            print("Could not load or parse Actors.json: \(error)")
        }
        return [] // Return empty on error to prevent repeated load attempts
    }()

    private static let _actorDefinitionsByNameStorage: [String: [String: Any]]? = {
        guard let definitions = _actorDefinitionsStorage else { return [:] }
        var byName: [String: [String: Any]] = [:]
        for actorDict in definitions {
            if let name = actorDict["name"] as? String {
                byName[name] = actorDict
            }
        }
        return byName
    }()

    // Publicly accessible static computed properties that use the lazy storage
    public static var actorDefinitions: [[String: Any]]? {
        return _actorDefinitionsStorage
    }

    public static var actorDefinitionsByName: [String: [String: Any]]? {
        return _actorDefinitionsByNameStorage
    }
    
    // No explicit loadActorDefinitions() needed due to lazy static properties.
    // Accessing actorDefinitions or actorDefinitionsByName will trigger the load.

    @objc public static func actorData(forIndex index: Int) -> [String: Any]? { // index is 1-based
        guard let definitions = self.actorDefinitions, index > 0, index <= definitions.count else {
            print("Error: actorData(forIndex:) index \(index) out of bounds (1-\(self.actorDefinitions?.count ?? 0)).")
            return nil
        }
        return definitions[index - 1] // Convert 1-based to 0-based
    }

    @objc public static func actorIndex(forName name: String) -> Int { // returns 1-based index or NSNotFound
        guard let definitionsByName = self.actorDefinitionsByName, let actorDict = definitionsByName[name], let id = actorDict["id"] as? Int else {
            return NSNotFound 
        }
        return id
    }

    // --- Instance Methods ---

    @objc public func spriteName() -> String? {
        guard let data = RWTActor.actorData(forIndex: self.actorIndex) else { return nil }
        return data["spriteBase"] as? String
    }

    @objc public func highlightedSpriteName() -> String? {
        guard let data = RWTActor.actorData(forIndex: self.actorIndex),
              let base = data["spriteBase"] as? String,
              let suffix = data["highlightedSuffix"] as? String else { return nil }
        return "\(base)\(suffix)"
    }

    // Corrected typo from blueHilightedSpriteNames
    @objc public func blueHighlightedSpriteName() -> String? {
        guard let data = RWTActor.actorData(forIndex: self.actorIndex),
              let base = data["spriteBase"] as? String,
              let suffix = data["blueSuffix"] as? String else { return nil }
        return "\(base)\(suffix)"
    }
    
    // The Objective-C instance method -(NSUInteger)actorIndex:(NSString *)actorName is omitted.
    // RWTLevel.m was already updated to use `[RWTActor actorIndexForName:]` which will now
    // map to the static Swift func `RWTActor.actorIndex(forName:)`.
    // The conversion from 1-based ID (from JSON) to 0-based index, if needed by specific
    // Obj-C clients, should be handled by those clients or via a specific Swift helper if required.
    // RWTLevel.m's createCurrentMovieActors was:
    // NSInteger actorIdFromName = [RWTActor actorIndexForName:selectedActorName]; // This is 1-based or NSNotFound
    // if (actorIdFromName != NSNotFound) { index = actorIdFromName; /* this index is 1-based */ }
    // This pattern is compatible with the new Swift static method.
}

// Helper for DispatchQueue.once behavior if needed, though lazy static vars are preferred.
// public extension DispatchQueue {
//    private static var _onceTracker = [String]()
//    static func once(token: String, block: () -> Void) {
//        objc_sync_enter(self)
//        defer { objc_sync_exit(self) }
//        if _onceTracker.contains(token) { return }
//        _onceTracker.append(token)
//        block()
//    }
// }
// Example usage: DispatchQueue.once(token: "com.example.loadActorDefinitions", block: { ... })
// But for this case, lazy static properties achieve the goal more simply.
